common/finalizer/finalizer.go

```diff
@@ -0,0 +1,160 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package finalizer
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/internal/goro"
+)
+
+var (
+	FinalizerAlreadyDoneErr = errors.New("finalizer already finalized")
+	FinalizerUnknownIdErr   = errors.New("finalizer callback not found")
+	FinalizerDuplicateIdErr = errors.New("finalizer callback already registered")
+)
+
+type Finalizer struct {
+	logger    log.Logger
+	mu        sync.Mutex
+	wg        sync.WaitGroup
+	finalized atomic.Bool
+	callbacks map[string]func(context.Context) error
+}
+
+func NewFinalizer(
+	logger log.Logger,
+) *Finalizer {
+	return &Finalizer{
+		logger:    logger,
+		callbacks: make(map[string]func(context.Context) error),
+	}
+}
+
+func (f *Finalizer) Register(
+	id string,
+	callback func(context.Context) error,
+) error {
+	if f.finalized.Load() {
+		return FinalizerAlreadyDoneErr
+	}
+
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	if _, ok := f.callbacks[id]; ok {
+		return FinalizerDuplicateIdErr
+	}
+	f.callbacks[id] = callback
+	return nil
+}
+
+func (f *Finalizer) Deregister(
+	id string,
+) (err error) {
+	if f.finalized.Load() {
+		return FinalizerAlreadyDoneErr
+	}
+
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	if _, ok := f.callbacks[id]; !ok {
+		return FinalizerUnknownIdErr
+	}
+	delete(f.callbacks, id)
+	return nil
+}
+
+func (f *Finalizer) Run(
+	pool *goro.AdaptivePool,
+	timeout time.Duration,
+) int32 {
+	if !f.finalized.CompareAndSwap(false, true) {
+		f.logger.Warn("finalizer skipped: called more than once")
+		return 0
+	}
+
+	if timeout == 0 {
+		f.logger.Warn("finalizer skipped: non-zero timeout")
+		return 0
+	}
+
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	totalCount := len(f.callbacks)
+	f.logger.Info("finalizer starting",
+		tag.NewInt("items", totalCount),
+		tag.NewDurationTag("timeout", timeout))
+
+	var completionCounter atomic.Int32
+	var wg sync.WaitGroup
+	wg.Add(totalCount)
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	go func() {
+		for _, cb := range f.callbacks {
+			func(callback func(context.Context) error) {
```

Comment:
Why do you need this func and another go routine? `pool.Do` is async, you should be able just call it for every `cb`.

Reply:
Ah, you're right. That's still left from before I added the pool. I'll remove it.

Reply:
Actually ... looking closer, wouldn't this trigger the for loop scoping issue if I remove it? https://go.dev/blog/loopvar-preview

Reply:
In go 1.22, it is on by default. So it should capture right `cb` on every iteration.

Reply:
Ah! I didn't catch that it was already enabled now. I thought it was still in preview. Time flies.